### ISSUE TITLE
`invalid-key` not detected for intersections that include `TypedDict`s

### ISSUE THREAD

**Issue by mudgett06** [astral-sh/ty#2591]
### Summary

## Summary                                                               
                                                                         
After narrowing a `TypedDict | None` type via `if not x: raise`, ty no   
longer detects invalid key access. The `invalid-key` rule works correctly when the variable is directly typed as the `TypedDict`.                   
                                                                         
## Reproduction                                                          
                                                                         
```python                                                                
from typing import TypedDict                                             
                                                                         
class Person(TypedDict):                                                 
    name: str                                                            
    age: int                                                             
                                                                         
def get_person() -> Person | None:                                       
    return None                                                          
                                                                         
def test() -> None:                                                      
    p = get_person()                                                     
    if not p:                                                            
        raise ValueError("No person")                                    
    # After narrowing, p should be Person - but invalid-key not triggered
    print(p["nonexistent"])  # No error (should be invalid-key)          
                                                                         
def test_direct(p: Person) -> None:                                      
    # Direct parameter works correctly                                   
    print(p["nonexistent"])  # Error: invalid-key ✓                      
                                                                         
```
### Expected Behavior                                                        
                                                                         
Both `test()` and `test_direct()` should emit `invalid-key` for                
`p["nonexistent"]`.                                                        
                                                                         
### Actual Behavior                                                          
                                                                         
Only `test_direct()` emits the error. After narrowing from `Person | None`,  
the `TypedDict` key information is lost.                                   
                                                                                                                 
                                                                         
Related                                                                  
                                                                         
Possibly related to TypedDict tracking issue #154                       

### Version

0.0.13

**Comment by AlexWaygood** [comment]
Thanks! The underlying cause here is that we infer a type of `Person & ~AlwaysFalsy` rather than simply `Person` after you've narrowed away the `None` type from the union. https://github.com/astral-sh/ruff/pull/22808 fixes the symptom you're experiencing here by improving our truthiness inference for `TypedDict`s with >=1 required key so that we understand that they're already subtypes of `~AlwaysFalsy`. But ideally we would also run this check on intersections that include `TypedDict` types too, so I don't think that that PR should close this issue.

**Comment by charliermarsh** [comment]
@AlexWaygood -- was this fixed by #22654?

**Comment by AlexWaygood** [comment]
> [@AlexWaygood](https://github.com/AlexWaygood) -- was this fixed by #22654?

Yes, looks like it -- that's awesome! We should add this as a test case -- it failed before that PR (https://github.com/astral-sh/ruff/pull/22808 was insufficient), but now passes on `main` following https://github.com/astral-sh/ruff/pull/22654 landing:

```py
from typing import TypedDict                                             
                                                                         
class Person(TypedDict, total=False):                                                 
    name: str                                                            
    age: int                                                             

def get_person() -> Person | None:                                       
    return None
                                                                         
def test() -> None:                                                      
    p = get_person()                                                     
    if not p:                                                            
        raise ValueError("No person")
    print(p["nonexistent"])  # error [invalid-key]
